src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs

```diff
@@ -644,6 +644,88 @@ protected override void ApplyAnalyzerReferenceRemoved(ProjectId projectId, Analy
+            {
+                var document = this.CurrentSolution.GetDocument(id);
+                // Only consider files that are mapped and that we are unable to apply changes to.
+                return document?.Services?.GetService<ISpanMappingService>() != null && document?.CanApplyChange() == false;
+            }
+        }
+
+        private void ApplyChangesForMappedDocuments(ProjectChanges projectChanges, ImmutableArray<DocumentId> changedMappedDocuments)
+        {
+            // Get the original text changes from all documents and call the span mapping service to get span mappings for the text changes.
+            var mappedSpansPerDocument = GetTextChangesAndMappedSpans(projectChanges, changedMappedDocuments);
+
+            // Create mapped text changes using the mapped spans and original text changes.
+            using var _ = ArrayBuilder<(string FileName, TextChange TextChange)>.GetInstance(out var mappedChangesBuilder);
+            foreach (var mappedSpans in mappedSpansPerDocument)
+            {
+                var newDoc = projectChanges.NewProject.GetDocument(mappedSpans.Key)!;
+                MapTextChanges(newDoc, mappedSpans.Value.MappedSpans, mappedSpans.Value.OriginalTextChanges, mappedChangesBuilder);
+            }
+
+            var mappedChanges = mappedChangesBuilder.ToImmutable();
+            if (mappedChanges.Any())
+            {
+                // Group the mapped text changes by file, then apply all mapped text changes for the file.
+                foreach (var changesForFile in mappedChanges.GroupBy(change => change.FileName))
+                {
+                    using var invisibleEditor = new InvisibleEditor(ServiceProvider.GlobalProvider, changesForFile.Key, GetHierarchy(projectChanges.ProjectId), needsSave: true, needsUndoDisabled: false);
+                    TextEditApplication.UpdateText(changesForFile.Select(change => change.TextChange).ToImmutableArray(), invisibleEditor.TextBuffer, EditOptions.None);
+                }
+            }
+
+            static Dictionary<DocumentId, (ImmutableArray<MappedSpanResult> MappedSpans, ImmutableArray<TextChange> OriginalTextChanges)> GetTextChangesAndMappedSpans(
+                ProjectChanges projectChanges, ImmutableArray<DocumentId> changedMappedDocuments)
+            {
+                return Task.Run(async () =>
+                {
+                    var mappedSpansPerDocument = new Dictionary<DocumentId, (ImmutableArray<MappedSpanResult> MappedSpans, ImmutableArray<TextChange> OriginalTextChanges)>();
+                    foreach (var changedDocumentId in changedMappedDocuments)
+                    {
+                        var oldDocument = projectChanges.OldProject.GetDocument(changedDocumentId)!;
+                        var newDocument = projectChanges.NewProject.GetDocument(changedDocumentId)!;
+
+                        var mappingService = newDocument.Services.GetService<ISpanMappingService>();
```

Comment:
```suggestion
                        var mappingService = newDocument.Services.GetRequiredService<ISpanMappingService>();
```

?

Reply:
I couldn't find an applicable GetRequiredService for document services, so will leave as-is for the moment,